modules/standard/WeakPointer.chpl

```diff
@@ -0,0 +1,334 @@
+/*
+ * Copyright 2020-2023 Hewlett Packard Enterprise Development LP
+ * Copyright 2004-2019 Cray Inc.
+ * Other additional copyright holders may be indicated within.
+ *
+ * The entirety of this work is licensed under the Apache License,
+ * Version 2.0 (the "License"); you may not use this file except
+ * in compliance with the License.
+ *
+ * You may obtain a copy of the License at
+ *
+ *     http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+
+
+
+/*
+This module contains the ``weak`` type, which is a smart pointer type designed
+to be used in tandem with :type:`~SharedObject.shared` objects.
+
+A ``weak`` provides a reference to a ``shared`` class object without
+requiring it to stay allocated. Such a pattern is useful for implementing graph
+or tree structures with bidirectional references, or for implementing cache-like
+data structures that maintain a list of objects but don't require them to stay
+allocated.
+
+A "strong" shared reference to the relevant class object can be obtained via
+the :proc:`~WeakPointer.weak.upgrade` method, or by casting the
+``weak`` to a ``shared t`` or a ``shared t?``. If the underlying object is
+not valid (i.e., its shared reference count has already dropped to zero
+causing it to be de-initialized) the upgrade attempt will fail.
+
+Weak pointers are implemented using task-safe reference counting.
+*/
+@unstable("The `weak` type is experimental; expect this API to change in the future")
+module WeakPointer {
+  use Errors, Atomics, ChapelBase, SharedObject;
+
+  record weak : writeSerializable {
+    /* The shared class type referenced by this pointer */
+    type classType;
+
+    pragma "owned"
+    var chpl_p: __primitive("to nilable class", _to_unmanaged(classType)); // instance pointer
+
+    pragma "owned"
+    var chpl_pn: unmanaged ReferenceCount?; // reference counter
+
+    // ---------------- Initializers ----------------
+
+    @chpldoc.nodoc
+    proc init(c : unmanaged) {
+      this.classType = c.type;
```

Comment:
Not that it matters since it's going to be a failure anyways (and I know it was this way before), but it seems like this field initialization is unnecessary.

Reply:
I think the compiler still complains about the type not being fully concrete without this line since it's a `type` field.